### PR TITLE
Convert sample app to ROS2

### PR DESCRIPTION
**Pull request by timrobotson**
Converts both robot and simulation to ROS2. I've made the following changes to the sample app:

- Removed all ros-navigation references and includes because we don't need it in this sample app, it was originally taken from cloudwatch, and it makes it a lot more complex. 
- Copied AudioData message from https://github.com/ros-drivers/audio_common/blob/master/audio_common_msgs/msg/AudioData.msg because `audio_common_msgs` hasn't been proted to ROS2 yet. 

text_input.py is working and integration_test is running but not passing yet. 

Requires https://github.com/aws-robotics/lex-ros2/pull/7 to be merged to work. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: robot_ws/src/voice_interaction_robot/scripts/text_input.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/scripts/text_input.py]: This is still `rospy.Publisher` is ros1 right?

File: robot_ws/src/voice_interaction_robot/scripts/audio_input.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/scripts/audio_input.py]: still has rospy

File: simulation_ws/src/turtlebot3_description_reduced_mesh/launch/spawn_turtlebot.launch.py
Comment (ryanewel) [inline simulation_ws/src/turtlebot3_description_reduced_mesh/launch/spawn_turtlebot.launch.py]: Shouldn't these files be coming from a shared package? A lot of sample apps need these reduced mesh turtlebot files

File: simulation_ws/src/voice_interaction_simulation/launch/spawn_balls.launch
Comment (ryanewel) [inline simulation_ws/src/voice_interaction_simulation/launch/spawn_balls.launch]: Why are we keeping this ros1 launch file?

**Review by ryanewel — CHANGES_REQUESTED** [review]
It looks like there is some ROS1 stuff still in this package.

File: robot_ws/src/voice_interaction_robot/launch/deploy_voice_interaction.launch.py
Comment (AAlon) [inline robot_ws/src/voice_interaction_robot/launch/deploy_voice_interaction.launch.py:32]: Might as well use os.path.join to build the entire path:
`os.path.join(get_package_share_directory('turtlebot3_bringup'), 'launch', 'turtlebot3_robot.launch.py')`

(Occurs in multiple places)

File: robot_ws/src/voice_interaction_robot/scripts/audio_input.py
Comment (AAlon) [inline robot_ws/src/voice_interaction_robot/scripts/audio_input.py]: Other than the rospy->rclpy issue, should we specify a suitable QoS profile? For this case I think QoSReliabilityPolicy makes sense: https://github.com/ros2/rclpy/blob/dashing/rclpy/rclpy/qos.py#L274

Same suggestion applies to all other publishers/subscribers we create too.

**Review by AAlon — COMMENTED** [review]
Can you rebase against the ros2 branch? CI should be fixed now.

File: robot_ws/src/voice_interaction_robot_msgs/msg/KeyValue.msg
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot_msgs/msg/KeyValue.msg]: Why can't we get this message type from lex_common_msgs? https://github.com/aws-robotics/lex-ros2/blob/master/lex_common_msgs/msg/KeyValue.msg
It looks like it's been released to dashing http://packages.ros.org/ros2/ubuntu/dists/bionic/main/binary-amd64/Packages

File: robot_ws/src/voice_interaction_robot_msgs/msg/AudioData.msg
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot_msgs/msg/AudioData.msg:1]: Once this gets merged we may want to open an issue to remind ourselves to delete this package once the messages are available in apt

File: robot_ws/.rosinstall
Comment (ryanewel) [inline robot_ws/.rosinstall:1]: We'll want to document somewhere that this should be removed once https://github.com/aws-robotics/lex-ros2/pull/7 has been released to apt

File: robot_ws/src/voice_interaction_robot/launch/voice_interaction.launch.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/launch/voice_interaction.launch.py]: Could you add a comment on why we're setting this env variable?

File: robot_ws/src/voice_interaction_robot/scripts/text_input.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/scripts/text_input.py]: It seems weird to put the spin statement in this function. Can this go in the main? Maybe you'd need a multithreaded executor?

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py:45]: Could you add units or some comment about what this variable represents?

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py]: Why remove this? There's `self.get_node_names()`
https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/node.py#L1546

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py]: Same comment on removing this. The functionality still exists in ROS2

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/voice_input.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/voice_input.py:29]: Don't need to change since this is the same in ROS1, but `p` isn't a super helpful variable name

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/voice_input.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/voice_input.py]: Don't need to change since this is the same in ROS1, but this feels like it could be made into a more generic function like `get_device_from_index(device_name)` or something that we call with `device_name="record"` or at least set a variable `device_name=record` at the start instead of using the string `record` all over the place.

File: robot_ws/src/voice_interaction_robot/launch/voice_interaction.launch.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/launch/voice_interaction.launch.py]: This launch argument is declared but never used. As is, the lex node just uses the default region in its config file no matter what is set here.

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py]: Should probably just delete this function

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py]: Remove commented out code

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py]: os.path.join?

File: robot_ws/src/voice_interaction_robot/launch/integration_test.launch.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/launch/integration_test.launch.py]: I think this causes the robot application to crash when launched in RoboMaker. We could add a parameter to disable this.

**Comment by AAlon** [comment]
(Rebased for the updated .travis.yml).

**Comment by AAlon** [comment]
@timrobotson Is this ready to  be merged once you have approvals? Have you had the chance to test both sim and robot workspaces? (are the integ tests passing?)

File: robot_ws/src/voice_interaction_robot/launch/voice_interaction.launch.py
Comment (timrobotson) [reply robot_ws/src/voice_interaction_robot/launch/voice_interaction.launch.py]: Ah true. Fixed, and submitted PR for lex_node to allow setting region and user_id via param https://github.com/aws-robotics/lex-ros2/pull/8/.

File: robot_ws/src/voice_interaction_robot_msgs/msg/AudioData.msg
Comment (timrobotson) [reply robot_ws/src/voice_interaction_robot_msgs/msg/AudioData.msg:1]: I don't think there's a need to use the other package, this is the only message type we're using from it and it's super simple.

File: robot_ws/.rosinstall
Comment (timrobotson) [reply robot_ws/.rosinstall:1]: Have noted it in my task list.

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (timrobotson) [reply robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py]: Thanks, didn't know about these.

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/voice_input.py
Comment (timrobotson) [reply robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/voice_input.py]: Good idea. Done.

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (timrobotson) [reply robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py]: Used https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/node.py#L1584 instead :)

File: robot_ws/src/voice_interaction_robot/scripts/audio_input.py
Comment (timrobotson) [reply robot_ws/src/voice_interaction_robot/scripts/audio_input.py]: I'll do this in a future PR and spend some time learning QoS profiles and trying them first.

**Comment by timrobotson** [comment]
All comments addressed, this should be ready to merge now. Will add proper QoS profiles in the future once I understand QoS better. Integration tests are passing and I've thoroughly tested robot_ws locally. Will try out sim + robot in RoboMaker. 

I haven't tested actual microphone input or speaker output completely yet. Will do that on a turtlebot.

File: robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py
Comment (ryanewel) [inline robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py:1]: License header should go before imports, right?

Comment (timrobotson) [reply robot_ws/src/voice_interaction_robot/src/voice_interaction_robot/test/integration_test.py:1]: Weird, pep8 formatting decided to do this :/

**Review by ryanewel — APPROVED** [review]

**Pull request merged**
